flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RedirectHandler.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.flink.runtime.rest.handler.util.HandlerRedirectUtils;
+import org.apache.flink.runtime.rest.handler.util.HandlerUtils;
+import org.apache.flink.runtime.rest.messages.ErrorResponseBody;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.OptionalConsumer;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandler;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+import org.apache.flink.shaded.netty4.io.netty.channel.SimpleChannelInboundHandler;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.router.KeepAliveWrite;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.router.Routed;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link SimpleChannelInboundHandler} which encapsulates the redirection logic for the
+ * REST endpoints.
+ *
+ * @param <T> type of the leader to retrieve
+ */
+@ChannelHandler.Sharable
+public abstract class RedirectHandler<T extends RestfulGateway> extends SimpleChannelInboundHandler<Routed> {
+
+	protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+	private final CompletableFuture<String> localAddressFuture;
+
+	protected final GatewayRetriever<T> leaderRetriever;
+
+	protected final Time timeout;
+
+	/** Whether the web service has https enabled. */
+	protected final boolean httpsEnabled;
+
+	private String localAddress;
+
+	protected RedirectHandler(
+			@Nonnull CompletableFuture<String> localAddressFuture,
+			@Nonnull GatewayRetriever<T> leaderRetriever,
+			@Nonnull Time timeout,
+			boolean httpsEnabled) {
+		this.localAddressFuture = Preconditions.checkNotNull(localAddressFuture);
+		this.leaderRetriever = Preconditions.checkNotNull(leaderRetriever);
+		this.timeout = Preconditions.checkNotNull(timeout);
+		this.httpsEnabled = httpsEnabled;
+		localAddress = null;
+	}
+
+	@Override
+	protected void channelRead0(
+		ChannelHandlerContext channelHandlerContext,
+		Routed routed) throws Exception {
+
+		try {
+			if (localAddressFuture.isDone()) {
+				if (localAddress == null) {
+					try {
+						localAddress = localAddressFuture.get(timeout.toMilliseconds(), TimeUnit.MILLISECONDS);
+					} catch (Exception e) {
+						logger.error("Could not obtain local address.", e);
+
+						HandlerUtils.sendErrorResponse(
+							channelHandlerContext,
+							routed.request(),
+							new ErrorResponseBody(new FlinkException("Could not obtain local address.", e)),
```

Comment:
I would prefer if we did not expose exception stacktraces directly in the API.

Reply:
Alright. Will remove them.